client/internal/dns/service_listener.go

```diff
@@ -140,55 +131,89 @@ func (s *serviceViaListener) setListenerStatus(running bool) {
 	s.listenerIsRunning = running
 }
 
-func (s *serviceViaListener) getFirstDNSListenerAvailable() (string, int, error) {
-	ips := []string{defaultIP, customIP}
+// evalListenAddress figure out the listen address for the DNS server
+// first check the 53 port availability on WG interface or lo, if not success
+// check the 5053 port on WG interface for eBPF, if not success
+// check the 5053 port availability on WG interface or lo without eBPF usage,
+func (s *serviceViaListener) evalListenAddress() (string, int, error) {
+	if s.customAddr != nil {
+		return s.customAddr.Addr().String(), int(s.customAddr.Port()), nil
+	}
+
+	ip, ok := s.testFreePort(defaultPort)
+	if ok {
+		return ip, defaultPort, nil
+	}
+
+	ebpfSrv, ok := s.tryToUseeBPF()
+	if ok {
+		s.ebpfService = ebpfSrv
+		return s.wgInterface.Address().IP.String(), customPort, nil
```

Comment:
we should probe this port or use a dynamic port with the code below:
```Go
	udpAddr := net.UDPAddrFromAddrPort(netip.MustParseAddrPort("0.0.0.0:0"))
	probeListener, err := net.ListenUDP("udp", udpAddr)
	var port uint16
	if err == nil {
		addPort := netip.MustParseAddrPort(probeListener.LocalAddr().String()) // might panic if address is incorrect
		port = addPort.Port()
		probeListener.Close()
		return s.wgInterface.Address().IP.String(),addPort.Port(), nil
	}
```

Reply:
In the tryToUseeBPF() function we test the port (I know would better outside but then the result source code more messy). I add the dynamic port.